### ISSUE TITLE
Test runner in VS Code extension

### ISSUE THREAD

**Issue by ffMathy** [oven-sh/bun#4824]
### What is the problem this feature would solve?

Bun tests can only be run from the CLI. 

### What is the feature you are proposing to solve the problem?

A VS Code extension could integrate the tests with the test explorer.

**Comment by alfu32** [comment]
maybe configure jest extension to run with bun

**Comment by ffMathy** [comment]
That's what #4562 proposes.

**Comment by Electroid** [comment]
We are working on a VSCode extension and it will have test runner support.

**Comment by ffMathy** [comment]
@Electroid great. Is there a way to track its progress somehow? Is the work being done on a particular branch?

And is it possible to provide some sort of ETA on when we can expect this? Eagerly waiting. Bun is awesome.

**Comment by Electroid** [comment]
> @Electroid great. Is there a way to track its progress somehow? Is the work being done on a particular branch?
> 
> And is it possible to provide some sort of ETA on when we can expect this? Eagerly waiting. Bun is awesome.

You can test it out [here](https://marketplace.visualstudio.com/items?itemName=oven.bun-vscode). Note that it does not have test support yet, but you can use the debugger. The code for the extension is in the `packages/bun-vscode` directory.

**Comment by orbiteleven** [comment]
@Electroid is there an issue tracking test support yet? Or is this the issue?

**Comment by ssalbdivad** [comment]
This would be a big deal for me in terms of adoption.

I tried bun out after updating some node dependencies and running into another ESM/CJS issue but would hate to give up the really clean test-runner integration.

**Comment by Kleywalker** [comment]
Is there a roadmap? Using test runner would increase productivity immensely! Especially `run test at cursor` and `rerun previous test` including keyboard shortcuts!

**Comment by ffMathy** [comment]
@Jarred-Sumner just one quick word on an ETA or if this is a priority in 2024? I don't see it on the roadmap. Is this intentional?

**Comment by kingkong404** [comment]
Is there anyway to achieve this functionality with the release of Bun 1.1 now that it has additional test framework support? (reconfigure any of the current vscode test runner plugins to use bun?)

**Comment by drk-mtr** [comment]
This would be a great addition.

> Is there anyway to achieve this functionality with the release of Bun 1.1 now that it has additional test framework support? (reconfigure any of the current vscode test runner plugins to use bun?)

Getting the [node:test runner](https://github.com/connor4312/nodejs-testing) library to discover tests is a two line change, but I wouldn't know how to integrate the reporting aspect.

![buntest](https://github.com/oven-sh/bun/assets/40315783/984070f9-a7ee-4496-92b6-ced018a2d885)

I wonder whether this could be done in two stages: for now, use this discovery to run a command in the terminal, then add proper integration with the [Testing API](https://code.visualstudio.com/api/extension-guides/testing) UI later. Though I guess this isn't a big step up from using the existing `Bun: Run File` functionality.

**Comment by ffMathy** [comment]
@drk-mtr I mean, even tests without coverage could be a fine start with a known limitation. Could you create a PR for this?

**Comment by drk-mtr** [comment]
Thanks but unfortunately the code I was playing with does nothing useful. It finds tests but it can't run them or report on the results.

**Comment by Kleywalker** [comment]
Are there any updates yet? Is anyone working on this? As projects getting larger running und debugging individual tests are not convenient but necessary. Keyboard shortcuts would speed up development a lot.

**Comment by ffMathy** [comment]
@Kleywalker doesn't look like it's being worked on. It's also not on Bun's updated roadmap 😦

**Comment by MaxMoldmann** [comment]
I would suggest that this single issue is one of the major user adoption hurdles. This needs a high priority.

**Comment by ffMathy** [comment]
I agree. Any official comment from the Bun team would be nice. I won't adopt Bun before this is ready.

**Comment by ducan-ne** [comment]
Looking forward for this 👀

**Comment by ffMathy** [comment]
Remember to give it a thumbs up reaction in the main post (not as a comment). I think that's how they prioritize.

**Comment by boehs** [comment]
https://github.com/oven-sh/bun/pull/13668

**Comment by meoyawn** [comment]
this is implemented in the official vscode extension https://marketplace.visualstudio.com/items?itemName=oven.bun-vscode

**Comment by drk-mtr** [comment]
Great to see this has been added, thanks!

I guess we need to decide whether (a) this issue is closed and a new issue is needed to track full integration with the Testing API (b) this issue will be considered closed once we have full integration.

By "full integration" I'm referring to 
1. Seeing tests in the Text Explorer i.e. press F1, then go to `Testing: Focus on Test Explorer View`
2. Seeing debug output in the UI as per https://code.visualstudio.com/updates/v1_93#_support-for-stack-traces-on-messages (as the node:test extension does as per this issue: https://github.com/connor4312/nodejs-testing/issues/37)

**Comment by kravetsone** [comment]
Full implementation will be like this: https://github.com/vitest-dev/vscode

![Image](https://github.com/user-attachments/assets/f9a14df2-687a-44f3-a032-1e72a1c767b1)

**Comment by Pckool** [comment]
Hey is this still in the works? Would really love to get the test explorer working properly with bun tests

**Comment by wzulfikar** [comment]
Tried some extensions to make this work ([Jest Test Explorer](https://marketplace.visualstudio.com/items?itemName=kavod-io.vscode-jest-test-adapter), [Test Explorer UI](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer)) but nothing works correctly. Took a stab to implement it myself based on the output of `bun test` and it seems to work for my codebases. Feel free to try and lmk if you have an issue.

- Github: github.com/wzulfikar/vscode-bun-test-explorer
- VSCode Marketplace: https://marketplace.visualstudio.com/items?itemName=wzulfikar.vscode-bun-test-explorer

Tested with:
- Cursor v0.49.4 (VSCode v1.96.2)
- VSCode v1.99.3
- Bun v1.2.12

Limitation:
- It uses output from `bun test`. I'm using bun v1.2.12, so if you're using bun version where the test output might differ, the extension may not work. If that happens, ,lmk and i'll try to update the [test output parser](https://github.com/wzulfikar/vscode-bun-test-explorer/blob/main/src/utils/parseBunTestOutput.ts).

https://github.com/user-attachments/assets/e6ad9774-e029-4550-ba74-cd4a3ef8a839

**Comment by wzulfikar** [comment]
Oh you are right. Will take a look the [bun extension code](https://github.com/oven-sh/bun/tree/main/packages/bun-vscode) and see if I can add the features there. Meanwhile will keep my extension available for people to try/give feedback.

**Comment by kravetsone** [comment]
> Oh you are right. Will take a look the [bun extension code](https://github.com/oven-sh/bun/tree/main/packages/bun-vscode) and see if I can add the features there. Meanwhile will keep my extension available for people to try/give feedback.
> 
> Update: cloned bun repo and managed to setup the dev workflow for the extension. Might be able to integrate the test explorer this weekend.

great work thank you!!

btw your repository source code is private

**Comment by kravetsone** [comment]
also take a look on

https://x.com/jarredsumner/status/1920393350957408687
https://github.com/jarred-sumner/bun-inspect-echo

**Comment by wzulfikar** [comment]
> Oh you are right. Will take a look the [bun extension code](https://github.com/oven-sh/bun/tree/main/packages/bun-vscode) and see if I can add the features there. Meanwhile will keep my extension available for people to try/give feedback.
> 
> Update: cloned bun repo and managed to setup the dev workflow for the extension. Might be able to integrate the test explorer this weekend.

Quick update: too the first stab to integrate the extension to bun repo (branch [here](https://github.com/wzulfikar/bun/tree/bun-vscode-test-explorer)), couldn't finish nicely. So for now I'll stick to using my extension but may do another attempt when I have some time. 

Longer story: added `BunTestController` to do initial test discovery and set up file watchers and refresh tests states when files change. Things become tricky when I start checking the details.

1. Test hierarchy is incorrect and tests are duplicates.

<img width=300 src=https://github.com/user-attachments/assets/62485f43-3503-4125-b18a-f01a45157346 />

2. Clicking a test in test explorer should trigger the file editor to open the test file at the correct line. Due to the duplicates above (i think), some of the tests are not triggering the file editor to open the test.
3. Wrong error overlay for failed tests

<img width=600 src=https://github.com/user-attachments/assets/f0cc5482-7e66-460c-aa04-44107c1b447d />

4. I'd want to show test output in the "Test Results" tab, but it didn't. It's just nice to have the output here i think, but not required. So maybe can be part of optional improvement or maybe it's just not needed.

<img width=600 src=https://github.com/user-attachments/assets/577eda5b-b4ae-43d5-9881-24d04eadb84e />

I might have better context when I do another attempt (unless someone beats me to it). But for now I'll leave this trace here.

**Comment by kravetsone** [comment]
Yeah I guess debugger echo should be used here

Also if something is missed I guess Jarred can help with it

**Comment by kravetsone** [comment]
> Would it be possible to also support test coverages similar to how Vitest extension on VSCode does it? On the explorer page, the folders and files show a percentage next to their names indicating the coverage amount.
> 
> <img width="362" height="166" alt="Image" src="https://github.com/user-attachments/assets/f765d99d-8fc4-4d2a-9489-4e984f80de17" />

Can u create a new issue for it?
